synapse/models/infotech.py

```diff
@@ -71,9 +65,211 @@ def getBaseModels():
+                    ('host', {'ptype': 'it:host'}),
+                    ('proc', {'ptype': 'it:exec:proc'}),
+                    ('time', {'ptype': 'time'}),
+
+                    ('exe', {'ptype': 'file:bytes', 'doc': 'The file containing code which retrieved the URL'}),
+                    ('ipv4', {'ptype': 'inet:ipv4', 'doc': 'The IPv4 address of the host during URL retrieval'}),
+                    ('ipv6', {'ptype': 'inet:ipv6', 'doc': 'The IPv6 address of the host during URL retrieval'}),
+                )),
+
+                ('it:exec:bind:tcp', {'ptype': 'guid'}, (
+                    ('port', {'ptype': 'inet:port', 'doc': 'The bound TCP port'}),
+
+                    ('ipv4', {'ptype': 'inet:ipv4', 'doc': 'The (optional) IPv4 specified to bind()'}),
+                    ('ipv6', {'ptype': 'inet:ipv6', 'doc': 'The (optional) IPv6 specified to bind()'}),
+                    ('host', {'ptype': 'it:host'}),
+                    ('proc', {'ptype': 'it:exec:proc'}),
+                    ('time', {'ptype': 'time'}),
+
+                    ('exe', {'ptype': 'file:bytes', 'doc': 'The file containing code which bound the listener'}),
+                )),
+
+                ('it:exec:bind:udp', {'ptype': 'guid'}, (
+                    ('port', {'ptype': 'inet:port', 'doc': 'The bound UDP port'}),
+
+                    ('ipv4', {'ptype': 'inet:ipv4', 'doc': 'The (optional) IPv4 specified to bind()'}),
+                    ('ipv6', {'ptype': 'inet:ipv6', 'doc': 'The (optional) IPv6 specified to bind()'}),
+                    ('host', {'ptype': 'it:host'}),
+                    ('proc', {'ptype': 'it:exec:proc'}),
+                    ('time', {'ptype': 'time'}),
+
+                    ('exe', {'ptype': 'file:bytes', 'doc': 'The file containing code which bound the listener'}),
+                )),
+
+                ('it:fs:file', {'ptype': 'guid'}, (
+                    ('host', {'ptype': 'it:host'}),
+
+                    ('path', {'ptype': 'file:path'}),
+                    ('path:dir', {'ptype': 'file:path', 'ro': 1}),
+                    ('path:ext', {'ptype': 'str:lwr', 'ro': 1}),
+                    ('path:base', {'ptype': 'file:base', 'ro': 1}),
+                    ('file', {'ptype': 'file:bytes'}),
+
+                    ('ctime', {'ptype': 'time', 'doc': 'File creation time'}),
+                    ('mtime', {'ptype': 'time', 'doc': 'File modification time'}),
+                    ('atime', {'ptype': 'time', 'doc': 'File access time'}),
+
+                    ('user', {'ptype': 'inet:user', 'doc': 'The owner of the file'}),
+                    ('group', {'ptype': 'inet:user', 'doc': 'The group owner of the file'}),
+                )),
+
+                # FIXME seed for hex file bytes
+                ('it:exec:file:add', {'ptype': 'guid'}, (
+                    ('host', {'ptype': 'it:host'}),
+                    ('path', {'ptype': 'file:path'}),
+                    ('path:dir', {'ptype': 'file:path', 'ro': 1}),
```

Comment:
Do these path subs have to be declared here? Won't they be added anyway?

Reply:
no, subs of a type will only be added to a node if they are declared.